File: src/main/java/app/tweditor/SaveFile.java
package app.tweditor;

import javax.swing.*;
import java.io.IOException;

public class SaveFile extends Thread
{
  private final ProgressDialog progressDialog;
  private final GameSession session;
  private final AppEnvironment environment;
  private boolean saveSuccessful = false;

  public SaveFile(ProgressDialog dialog, GameSession session, AppEnvironment environment)
  {
    this.progressDialog = dialog;
    this.session = session;
    this.environment = environment;
  }

  public void run()
  {
    try
    {
      this.session.getDatabase().save();
      this.progressDialog.updateProgress(15);

      ResourceEntry resourceEntry = new ResourceEntry("module.ifo", this.session.getDatabaseFile());
      this.session.getModDatabase().addEntry(resourceEntry);
      this.session.getModDatabase().save();
      this.progressDialog.updateProgress(30);

      ResourceDatabase modDatabase = new ResourceDatabase(this.session.getModDatabase().getPath());
      modDatabase.load();
      this.session.setModDatabase(modDatabase);
      this.progressDialog.updateProgress(45);

      this.session.getSaveDatabase().addEntry(this.session.getModName(), this.session.getModFile());
      this.progressDialog.updateProgress(60);

      this.session.getPlayerDatabase().save();
      this.session.getSaveDatabase().addEntry(this.session.getPlayerName(), this.session.getPlayerFile());
      this.progressDialog.updateProgress(70);

      this.session.getSmmDatabase().save();
      this.session.getSaveDatabase().addEntry(this.session.getSmmName(), this.session.getSmmFile());
      this.progressDialog.updateProgress(80);

      this.session.getSaveDatabase().save();
      this.progressDialog.updateProgress(90);

      SaveDatabase saveDatabase = new SaveDatabase(this.environment, this.session.getSaveDatabase().getPath());
      saveDatabase.load();
      this.session.setSaveDatabase(saveDatabase);

      this.progressDialog.updateProgress(100);

      this.saveSuccessful = true;
    } catch (DBException exc) {
      Main.logException("Unable to update save database", exc);
    } catch (IOException exc) {
      Main.logException("Unable to save file", exc);
    } catch (Throwable exc) {
      Main.logException("Exception while saving file", exc);
    }

    SwingUtilities.invokeLater(() ->
            SaveFile.this.progressDialog.closeDialog(SaveFile.this.saveSuccessful));
  }
}

File: src/main/kotlin/app/tweditor/SaveFile.kt
package app.tweditor

import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class SaveFile private constructor(
    private val session: GameSession,
    private val environment: AppEnvironment,
    private val targetFile: File?,
    private val reportProgress: (Int) -> Unit,
    private val complete: (Boolean) -> Unit,
    private val reportError: (String, Throwable) -> Unit,
    @Suppress("UNUSED_PARAMETER") private val callbackMode: Boolean
) : Thread() {
    /** Headless/asynchronous seam used by the Compose shell and file-workflow tests. */
    constructor(
        session: GameSession,
        environment: AppEnvironment,
        targetFile: File? = null,
        onProgress: (Int) -> Unit,
        onComplete: (Boolean) -> Unit,
        onError: (String, Throwable) -> Unit = { text, exc -> Main.logException(text, exc) }
    ) : this(session, environment, targetFile, onProgress, onComplete, onError, true)

    private var saveSuccessful = false

    override fun run() {
        try {
            if (targetFile != null) {
                SavePipeline.rebindToCopy(session, environment, targetFile)
            }

            session.database!!.save()
            reportProgress(15)

            val resourceEntry = ResourceEntry("module.ifo", session.databaseFile)
            session.modDatabase!!.addEntry(resourceEntry)
            session.modDatabase!!.save()
            reportProgress(30)

            val modDatabase = ResourceDatabase(session.modDatabase!!.getPath())
            modDatabase.load()
            session.modDatabase = modDatabase
            reportProgress(45)

            session.saveDatabase!!.addEntry(session.getModName()!!, session.modFile)
            reportProgress(60)

            session.playerDatabase!!.save()
            session.saveDatabase!!.addEntry(session.getPlayerName()!!, session.playerFile)
            reportProgress(70)

            session.smmDatabase!!.save()
            session.saveDatabase!!.addEntry(session.getSmmName()!!, session.smmFile)
            reportProgress(80)

            if (session.isJournalDirty()) {
                FileOutputStream(session.questDatabaseFile).use { out ->
                    session.getQuestDatabase()!!.save(out)
                }
                session.saveDatabase!!.addEntry(session.getQuestDBName()!! + ".qdb", session.questDatabaseFile)
            }
            reportProgress(85)

            session.writeSave()
            reportProgress(90)

            val saveDatabase = SaveDatabase(environment, session.saveDatabase!!.getPath())
            saveDatabase.load()
            session.saveDatabase = saveDatabase

            reportProgress(100)

            this.saveSuccessful = true
        } catch (exc: DBException) {
            reportError("Unable to update save database", exc)
        } catch (exc: IOException) {
            reportError("Unable to save file", exc)
        } catch (exc: Throwable) {
            reportError("Exception while saving file", exc)
        }

        complete(saveSuccessful)
    }
}
